### ISSUE TITLE
Choose more useful default diff ranges for the PR author when there are no provisional revisions

### ISSUE THREAD

**Issue by erikwright** [Reviewable/Reviewable#329]
Context:
1. I authored a PR.
2. Alex commented on some removed code while also marking all files reviewed.
3. I responded, without adding any new commits.

When Alex returned, he was confused because the comment thread appeared without any context. Reviewable selected the latest revision (empty diff range) to display. The code change that the comment applied to was gone. Since the change was a deletion, there was nothing to indicate the comment's context.

Suggestion:

Initialize to the same diff range that was last viewed (at least if there are no new commits).

**Comment by pkaminski** [comment]
I thought about this a bit more, and perhaps the "right" thing to do is to switch each file to a diff range that minimizes context errors in its unresolved comments.  Given that Reviewable automatically switches to the latest changes if there's new revisions (and I think this behavior should remain), it could be confusing if sometimes it went to your last selection and sometimes picked a new range.  So maybe it's better to always pick the "most useful" range on page load.

**Comment by erikwright** [comment]
You're the boss, but I think any default view that shows different files at
different revisions will be confusing. After thinking about it, I suppose
that can already happen now (if I mark only part of a PR as reviewed), and
that seems unexpected and likely to lead to confusion.

The more I think about it, the more I think that the right thing is to
always restore the previous diff range (or at least the previous lower
bound) and offer a single-click to go to "unreviewed diffs".

But again, you're the boss here. Knowing about the little corner icon
solves the most pressing usability issues for me.

**Comment by pkaminski** [comment]
Right, automatically selecting different diff ranges for the files already happens quite often.  I haven't yet heard any complaints about it so it's probably not *too* confusing.  The single button to go to the "next unreviewed diffs" (or, in general, "Reviewable's suggest diff set") already exists, it's the middle button in the Changes box.  Currently, Reviewable essentially clicks it for you at page load; what you're proposing is always loading the previous view, and letting the user click it.  Both approaches are consistent, and it's not clear to me yet where the optimum trade-off between "least surprise" and "most usefulness" lies in this case -- I'll need to think about it some more.  Either way, selecting the best ranges for viewing comments (possibly with the right bound forced to the latest revision?) seems like a useful thing to do.

**Comment by erikwright** [comment]
PREFIX: This is all opinion and conjecture.

My opinion is that the reviewer probably left the view in the appropriate
state for whatever discussion they were engaged in, so just leaving things
like that is actually quite likely to give them what they want.

Furthermore, I think it's incredibly easy for Reviewable to offer a
one-click "show me new changes" button, and the user is very likely to
click that when they want it.

But it's harder (or at least, more confusing and less discoverable) to
offer a "restore to previous state" button. The user is only likely to
realize they want this after attempting to respond to one or more comments,
and the mental leap is higher for realizing that the way to get their
context is to push this button.

I would also push back on this:

> automatically selecting different diff ranges for the files already
happens quite often. I haven't yet heard any complaints about it so it's
probably not *too* confusing.

This only happens if I mark only a subset of the files as reviewed. Maybe
you measure how often that happens, but I doubt it's the norm.

**Comment by pkaminski** [comment]
I posted a poll, to inform the decision with a modicum of data: https://twitter.com/reviewableio/status/740265469219852288

**Comment by pkaminski** [comment]
We decided that we won't save and restore the last-viewed diff range, as it would get confusing when files were added to the PR since the last save point, and also because nobody else has asked for this in years.

We'll also keep the current behavior of diffing the provisional revisions for the PR author by default, but add a new feature that lets users pick a different default from among a set of standard diffs.  If we get an idea later for a more useful default diff view for the author we can add it to this set.
